File: tests/pages/TodoPage.ts
import { Locator, Page, expect } from '@playwright/test';

export class TodoPage {
  readonly page: Page;
  readonly heading: Locator;
  readonly todoInput: Locator;
  readonly addButton: Locator;
  readonly todoItems: Locator;
  readonly itemsCount: Locator;

  constructor(page: Page) {
    this.page = page;
    // getByRole: Semantic heading identification
    this.heading = page.getByRole('heading', { name: 'Todo App' });
    // getByLabel: Finds input via associated <label> (best for forms)
    this.todoInput = page.getByLabel('New todo');
    // getByRole: Targets the specific functional button
    this.addButton = page.getByRole('button', { name: 'Add' });
    // getByTestId: Reliable selection for repeating list items
    this.todoItems = page.getByTestId('todo-item');
    // getByRole: Targets element with role="status"
    this.itemsCount = page.getByRole('status');
  }

  async goto() {
    await this.page.goto('/');
  }

  async addTodo(title: string) {
    await this.todoInput.fill(title);
    await this.addButton.click();
  }

  async completeTodo(title: string) {
    // getByRole with dynamic name based on ARIA label in index.html
    await this.page.getByRole('button', { name: `Complete: ${title}` }).click();
  }

  async deleteTodo(title: string) {
    // getByRole with dynamic name based on ARIA label in index.html
    await this.page.getByRole('button', { name: `Delete: ${title}` }).click();
  }

  /**
   * Cleans up state via API for maximum speed and reliability.
   */
  async resetViaApi() {
    await this.page.request.delete('/api/todos');
  }
}

File: tests-python/pages/todo_page.py
import re
from playwright.sync_api import Page, Locator


class TodoPage:
    def __init__(self, page: Page) -> None:
        self.page = page

        # get_by_label: finds the input via <label for="todo-input"> — best for forms
        self.new_todo_input: Locator = page.get_by_label("New todo")

        # get_by_role: button with accessible name "Add"
        self.add_button: Locator = page.get_by_role("button", name="Add")

        # get_by_test_id: data-testid="todo-item" — reliable for repeated rows
        self.todo_items: Locator = page.get_by_test_id("todo-item")

        # get_by_role "status": the counter paragraph with role="status" and aria-live
        self.item_count: Locator = page.get_by_role("status")

    # ── Navigation ───────────────────────────────────────────────────────────

    def goto(self) -> None:
        self.page.goto("http://localhost:8000")

    # ── Actions ──────────────────────────────────────────────────────────────

    def add_todo(self, title: str) -> None:
        self.new_todo_input.fill(title)
        self.add_button.click()

    def complete_todo(self, title: str) -> None:
        # Scoped to the matching row to avoid ambiguity when titles share substrings
        self.todo_items.filter(has_text=title).get_by_role(
            "button", name=re.compile(f"Complete: {re.escape(title)}")
        ).click()

    def delete_todo(self, title: str) -> None:
        self.todo_items.filter(has_text=title).get_by_role(
            "button", name=re.compile(f"Delete: {re.escape(title)}")
        ).click()

    # ── Test helpers ─────────────────────────────────────────────────────────

    def reset_via_api(self) -> None:
        """Clears all todos via REST API — faster and more reliable than UI clicks."""
        self.page.request.delete("http://localhost:8000/api/todos")
